sound/soc/sof/topology.c

```diff
@@ -2130,7 +2130,7 @@ static int sof_get_control_data(struct snd_soc_component *scomp,
 	for (i = 0; i < widget->num_kcontrols; i++) {
 		kc = &widget->kcontrol_news[i];
 
-		switch (widget->dobj.widget.kcontrol_type) {
```

Comment:
well something is really odd here. If we want to 'enable multiple type controls for single widgets', then this field should be removed completely.

Of why not instead make kcontrol_type an array. e.g.
````c
		switch (widget->dobj.widget.kcontrol_type[i]) {	
````

and don't change the kcontrols themselves.

Reply:
hmm I've changed this already, I think I've not updated the correct version, sorry about this, will push new version

Reply:
oh sorry will take it back. was confused by the diff...

So yes was thinking about the array, but what is would be the size of the array? anyway we are having the individual kcontrol structs, so why not have types in those? Don't have to worry/check array bounds or dynamic memory allocation..

Now thinking it, maybe the use of kcontrol_new struct is not the best then. I think we have "somewhere" present 3 different structs: 1) kcontrol_new, 2) actual alsa kcontrol (created from 1), and 3) sof_control. So some of those should/could be used here.

And I still insist that it would be good to have the type in the control struct, not in widget :)

Reply:
@ranj063 I think you are atm the best kcontrol expert on this planet, what do you think?

Reply:
sorry im not getting the question here. or has the fix already been pushed and I'm missing context?